src/relay/context.ts

```diff
@@ -157,22 +157,37 @@ class RelayContext extends EventEmitter {
         this._sourcePromise = this._stream.source.next()
       }
 
+      this.verbose(`FLOW: relay_incoming: started loop`)
       while (true) {
+        this.verbose(`FLOW: relay_incoming: new loop iteration`)
+
         const promises: Promise<Stream['source'] | StreamResult>[] = []
+        let resolvedPromiseName
```

Comment:
by using the promise name, we can probably get rid of ~20% of the state management complexity